### ISSUE TITLE
Support for nvcc?

### ISSUE THREAD

**Issue by teonnik** [build2/build2#3]
Is support for `nvcc` or other GPU / CUDA compilers available / planned?

**Comment by optomux** [comment]
Isn't nvcc just a thin wrapper for cl.exe?

**Comment by optomux** [comment]
Right.. so how to make it work with nvcc?

**Comment by boris-kolpackov** [comment]
Someone needs to write a `build2` rule for it. The best person to do that would be someone who has experience with nvcc (ideally expert-level) and a burning need for it (i.e., someone who has a real use-case on which we can test things). From our side we would be happy to provide guidance when it comes to the `build2` part of it (i.e., the person doesn't need to be a `build2` expert).

**Comment by teonnik** [comment]
If you want to do CUDA programming, clang supports it now : https://llvm.org/docs/CompileCudaWithLLVM.html . You may be able to get it working with build2. I haven't tried. 

I don't have an expert level knowledge of nvcc but I will be happy to give it a try once I have more time on my hands. (not the next 2 weeks)

**Comment by boris-kolpackov** [comment]
Interesting, thanks for the link. I skimmed through the documentation and I think it should be pretty straightforward to use Clang for CUDA compilation. We may want to do something for the `.cu` extension and `-x cuda` but that should be failrly minor. Other than that it looks like it should just work. Do you have any sense of how widely this (Clang/CUDA) is used compared to nvcc?

**Comment by teonnik** [comment]
Most people still use nvcc. At the last GPU Hackathon in CSCS, Switzerland, there was only one team using it (out of 10 or so). The HPC space (where I work), moves rather slowly and is riddled with proprietary "solutions". Unfortunately a clang toolchain is not provided by default on many supercomputers. 

A major advantage of clang cuda compared to nvcc is that nvcc often gives false positives for modern c++ libraries. Also, nvcc relies on a host compiler (gcc or clang) to compile the non-CUDA bits of the application. The clang CUDA compilation model appears to be a lot simpler (at least on the surface) since clang compiles C++ code anyway.

I don't know what the situation is in other domains.

**Comment by optomux** [comment]
Here is an example of nvcc usage. 
https://gist.github.com/misaka-10032/1a09d57e88e7697798d2b55a9e0f6428

**Comment by boris-kolpackov** [comment]
Thanks for the link. It would be also interesting to see a `makefile` that implements header dependency extraction. Do you have one of th0se? Since nvcc is based on Clang, it should support the `-M*` option family but it would be good to know whether it actually works in the light of that split compilation mode described in the Clang doc above.

**Comment by optomux** [comment]
You mean this?
https://stackoverflow.com/questions/19164974/generate-header-dependencies-for-cuda-files

**Comment by boris-kolpackov** [comment]
Yes, from that answer, it looks like nvcc only supports a subset (really just one) of `-M*` options. I think it's pretty clear that nvcc will require a custom `build2` rules and will only provide a subset of a normal C/C++ compiler functionality (like no support for auto-generated headers).

**Comment by asartori86** [comment]
Hi, any updates on this?

**Comment by boris-kolpackov** [comment]
I've had someone approach me at CppCon with interested in this but it doesn't seem to have lead to anything. The good news is that we now have support for external build system modules so implementing this outside of the `build2` core should be much easier. I am happy to help with the `build2` side of things but we really need someone with CUDA experience to champion this.

**Comment by asartori86** [comment]
@boris-kolpackov, I would be glad to give this a try. Could you point me where should I start from?

**Comment by boris-kolpackov** [comment]
I think as the first step it would be helpful to understand what's the current state and the future of the NVCC and Clang approaches. 

At least on the surface Clang looks like a much saner way but I did a bit of googling and it seems NVCC is still the predominantly used approach. Also, the [Clang's CUDA page](https://llvm.org/docs/CompileCudaWithLLVM.html) haven't been updated for a while which gives the impression that the whole idea is abandoned.

I also found this [CUDA compilers](https://gist.github.com/ax3l/9489132) with some relevant information.

**Comment by asartori86** [comment]
@boris-kolpackov as far as I know in the HPC world, `nvcc` is the de facto standard compiler to build Cuda applications. You should support it. Besides, along with clang-way, there are also the `pgi` compilers that could/should be added (`pragma acc` can be handy and powerful). However, if I were you, I would start supporting `nvcc` first, though.

**Comment by boris-kolpackov** [comment]
Yes, that's the impression I've got as well. I guess it's decided then, NVCC it is.

I think the next logical step would be to figure out how we are going to support the NVCC compilation model in `build2`. The only description of this model that I've seen so far is on the [Clang's CUDA page](https://llvm.org/docs/CompileCudaWithLLVM.html#compilation-models). Is this still accurate?

Generally, with `build2` we will probably want to handle as much of the under the hood mechanics as possible if we stand a chance of reusing the existing C++ compilation rules for host compilation (which I think will be very beneficial; I can elaborate why that is if needed). So the question for the CUDA experts is whether it's possible to decompose all the steps that NVCC performs under the hood (compile for device architectures, merge the result into a single "header", compile for host, etc) and perform them ourselves?

**Comment by boris-kolpackov** [comment]
FYI, some interesting feedback from one of the Clang CUDA developers in comments to the [CUDA compilers](https://gist.github.com/ax3l/9489132) page.
